threadpoolctl.py

```diff
@@ -315,6 +279,76 @@ def _set_threadpool_limits(self):
                 lib_controller.set_num_threads(num_threads)
 
 
+class _threadpool_limits_decorator(_threadpool_limits, ContextDecorator):
+    """Same as _threadpool_limits but to be used as a decorator"""
+
+    def __init__(self, controller, *, limits=None, user_api=None):
+        self._limits, self._user_api, self._prefixes = self._check_params(
+            limits, user_api
+        )
+        self._controller = controller
+
+    def __enter__(self):
+        # we need to set the limits here and not in the __init__ because we want the
+        # limits to be set when calling the decorated function, not when creating the
+        # decorator.
+        self._set_threadpool_limits()
+        return self
+
+
+@_format_docstring(
+    USER_APIS=", ".join(f'"{api}"' for api in _ALL_USER_APIS),
+    BLAS_LIBS=", ".join(_ALL_BLAS_LIBRARIES),
+    OPENMP_LIBS=", ".join(_ALL_OPENMP_LIBRARIES),
+)
+class threadpool_limits(_threadpool_limits):
+    """Change the maximal number of threads that can be used in thread pools.
+
+    This object can be used either as a callable (the construction of this object
+    limits the number of threads), as a context manager in a `with` block to
+    automatically restore the original state of the controlled libraries when exiting
+    the block, or as a decorator through its `wrap` method.
+
+    Set the maximal number of threads that can be used in thread pools used in
+    the supported libraries to `limit`. This function works for libraries that
+    are already loaded in the interpreter and can be changed dynamically.
+
```

Comment:
Maybe we could add a disclaimer to the docstrings of the public method to make it explicit that even when limits are set using the context manager or function decorator APIs, the effect is global on whole Python process: the is no thread level isolation as the underlying libraries do not offer thread-local APIs to configure the number to use in nested parallel calls.

See: https://github.com/joblib/threadpoolctl/issues/69#issuecomment-707136187

Reply:
The issue in the notebook from this comment is that the limits were set in the __init__ so creating the decorator would set the limits right away for all further blas calls.

But I agree that we can add that disclaimer anyway

Reply:
Done